### DISCUSSION TITLE
Writing to and reading from scalar aux variable

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#26806] [Q&A General] [answered]
Hello,

I'm calculating a scalar value that is constant over my whole domain that I'd like to store in an scalar aux variable

```
[scalar]
  family = SCALAR
  order = FIRST
[]
```

**Write**
Header-File
`MooseVariableScalar * _scalar_var;`
Constructor
`_scalar_var(&_subproblem.getScalarVariable(_tid, "var_name")`
Method
`_scalar_var->setValues(value);`

**Read**
Header-File
`MooseVariableScalar * _scalar_var;`
Constructor
`_scalar_var(&_subproblem.getScalarVariable(_tid, "var_name")`
Method
`Real value = _scalar_var->sln()[0]`

This code is working fine. My problem lies in the MOOSE Output which isn't updated throughout the simulation and stays 0, even though I can pass my scalar value using the scalar aux variable between classes. So my question is: How can I write the value from my solution vector to wherever MOOSE reads the output values from?

**Comment by GiudGiud** [comment]
Hello

Are you using a Scalar Auxkernel to set this variable? This would be the cannon way instead of setting it directly. 

This would also let you look at the timing of execution between output and ths scalar auxkernel

Guillaume

**Reply by Flolaffel** [reply]
I'm using a UserObject to set the variable. And for clarity reasons, I'd like to keep it in a UserObject, if that's possible.

**Reply by GiudGiud** [reply]
Which output is giving you the wrong value? The console output?

When are you executing these operations? What execute_on flag is passed to the user object?

**Reply by GiudGiud** [reply]
Also is a first order scalar variable one or two DoFs? I don't recall
And how many do you need?

**Reply by Flolaffel** [reply]
> The console output?

Yes, the console output doesn't match the values the objects are passing to each other.

> What execute_on flag is passed to the user object?

The UserObjects are executed at `TIMESTEP_END` and `post_aux`.

> Also is a first order scalar variable one or two DoFs?

First order

> And how many do you need?

How many scalar aux variables? Number varies based on simulation but will (for now) be 2 max.

**Reply by GiudGiud** [reply, marked as answer]
Can you use `Debug/show_execution_order=ALWAYS` to make sure the UOs are executed before the console prints?

If it s not an ordering issue, then the 'problem' happens in TableOutput::outputScalarVariable
where the scalar variable is reinited to be obtained  well-updated from the solution vector

I think what you are missing is a call to insert the value in the solution vector after setting the value.
```
MooseVariableScalar::insert(NumericVector<Number> & soln)
```
iirc you can get the solution vector from the nonlinear system which you can get from the problem

**Reply by Flolaffel** [reply]
That was it, thanks! I added the following lines to insert the value into the solution vector:

```
_scalar_var->insert(_scalar_var->sys().solution());
_scalar_var->sys().solution().close();
```

However I don't quite understand what exactly is happening here yet. Could please reiterate what is the difference between `setValues()` and `insert()`? I thought `setValues()` already sets the values of the solution vector?

**Reply by GiudGiud** [reply]
I think the call to close here at least is important to synchronize the solution vector. 

The explanation for both routines should be in the doxygen for moose
